### ISSUE TITLE
Auto upload Swagger files to SwaggerHub

### ISSUE THREAD

**Issue by ecristobal** [springfox/springfox#2424]
Good evening everybody!

I'm working on a project in my company in which we are using Springfox to generate automatically Swagger descriptors for our Spring services, and also we are using SwaggerHub to publish those files in order to give our developers a kind of developer portal where they can find all the services that we have published and their Swagger file to generate a client automatically.

Now we have only a few services, so we manually upload these files to SwaggerHub, but we find that more often than we want people forget to upload them, what causes a misalignment between what is published and their documentation on SwaggerHub. That's the reason why we thought that it would be great to have the possibility that, once the Swagger descriptor were autogenerated, auto upload it to SwaggerHub to avoid the problems caused by a manual upload that not always happen.

We have seen that SwagerHub has a REST API that could be used to perform that upload, so we think that it could be a great new feature to implement. We have made a fork of your code just to take a look at how we could implement that functionality, but without having taken a deep look on it maybe creating one Maven artifact and implementing a Strategy desing pattern will be enough, leaving on one of the core artifacts an interface that should be implemented into the new Maven artifact and gives support to auto upload to SwaggerHub, and being possible to implement this interface in other Maven artifacts to auto upload into other providers.

We have been looking for similar features but we haven't found any helpful resource. Anyway, don't hesitate to give any feedback about this idea which we're sure will be very helpful.

Best regards.

**Comment by dilipkrish** [comment]
@ecristobal This is a great idea. However I don't believe it belong in the core library. Not everyone uses swagger-hub and it feels weird for a service to auto-upload artifacts/contracts etc. without interaction from a user.

**Comment by ecristobal** [comment]
@dilipkrish I think I didn’t explain well my proposal on my previous post, because what I propose is that the core library just implements the extension point (if it doesn’t exists yet) where the SwaggerHub extension artifact will plug-in. This allows the possibility that, if someone wants to store their Swagger files into any other central store (some people could be thinking about Confluence, for example), they have one point in which they could plug in and perform the upload of their files.

Regarding what you say about feeling weird to auto-upload Swagger files without user interaction, think about it about closing the gap to get full automation on service descriptor generation: I think (and many other people agree with me) it will be quite powerful to automate all the process of creation and publishing of Swagger files, as it will ensure that at any time we will have on a central store the descriptors of the services running in our systems without user interaction.

Best regards!

**Comment by dilipkrish** [comment]
I agree it is useful to reduce the friction, my reservation is only in regards to the scope of this library. It could be achieved by an external tool, either a build step using maven/gradle or an external tool that runs on a schedule, or via a pull mechanism in swagger-hub (there might already be something that does that).

**Comment by ecristobal** [comment]
AFAIK SwaggerHub only allows file pushing by their web console and by their REST API, and I found the swaggerhub-maven-plugin (https://github.com/swagger-api/swaggerhub-maven-plugin) that can upload Swagger files to the site. This tool or any other requires that Swagger files should be generated on compile time, so Maven could grab and upload them. While this is not bad if you have automated compile and deployments because you configure once and use it many times, I think that the really powerful approach is yours, in which Springfox creates the Swagger files at runtime, on application startup, and to close the circle this generated Swagger auto-uploads to the project's "source of truth" (in this case, SwaggerHub), allowing everybody to forget to keep updated the APIs, as they update on every application start up, allowing an approach in which the library manages all the lifecycle of the Swagger file.
